File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    int len(ListNode* head){
        int count = 0;
        
        ListNode* temp = head;
        
        while(temp!=NULL){
            temp = temp->next;
            count++;
        }
        
        return count;
    }
    
    ListNode* reverseKhelper(ListNode* head,int k ,int l){
        
        if(l<k){
            return head;
            
        }
        
        ListNode* prev= NULL;
        ListNode* curr = head;
        ListNode* nextptr;
        
        int count = 0;
        
        
        while(curr!=NULL && count<k){
            nextptr = curr->next;
            curr->next = prev;
            prev=curr;
            curr = nextptr;
            count++;
        }
        
        if(nextptr!=NULL){
               head->next = reverseKhelper(nextptr,k,l-k);
        }
        
        return prev;  // prev wil be our new head
     
    
        
    }
    
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        
        int l = len(head);
        return reverseKhelper(head,k,l);
        
    }
};

File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    def get_length(self,head):
        temp = head
        count = 0
        while(temp):
            count+=1
            temp = temp.next
        return count
    
    def helper(self,head,l,k):
        
        if(l<k):
            return head
        
        curr=head
        prev = None
        count = 0
        nxt = None
        while(curr is not None and count<k):
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
            count+=1
        
        if(nxt):
            head.next = self.helper(nxt,l-k,k)
        
        return prev
        
            
        
        
    
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        l = self.get_length(head)
        return self.helper(head,l,k);
        